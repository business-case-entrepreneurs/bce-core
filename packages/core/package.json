{
  "name": "@bcase/core",
  "version": "0.0.27",
  "description": "",
  "license": "UNLICENSED",
  "author": {
    "name": "BCE Solutions",
    "email": "solutions@bcase.nl"
  },
  "files": [
    "dist",
    "loader",
    "scss"
  ],
  "main": "dist/index.js",
  "unpkg": "dist/bce/bce.js",
  "module": "dist/index.mjs",
  "types": "dist/types/types.d.ts",
  "scripts": {
    "build": "stencil build --docs-json ./dist/docs.json",
    "clean": "rimraf .stencil dist loader www src/components.d.ts",
    "serve": "stencil build --watch --serve --docs-json ./dist/docs.json"
  },
  "devDependencies": {
    "@fortawesome/fontawesome-svg-core": "^1.2.21",
    "@fortawesome/free-regular-svg-icons": "^5.10.2",
    "@fortawesome/free-solid-svg-icons": "^5.10.2",
    "@stencil/core": "^1.3.2",
    "@stencil/sass": "^1.0.1",
    "codeflask": "^1.4.1",
    "popper.js": "^1.15.0",
    "rimraf": "^2.6.3",
    "typescript": "^3.6.3"
  },
  "publishConfig": {
    "access": "public"
  },
  "collection": "dist/collection/collection-manifest.json",
  "collection:main": "dist/collection/index.js",
  "es2015": "dist/esm/index.mjs",
  "es2017": "dist/esm/index.mjs",
  "sass": "scss/index.scss"
}
