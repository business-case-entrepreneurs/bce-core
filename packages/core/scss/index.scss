@import './breakpoint';
@import './color';
@import './elevation';
