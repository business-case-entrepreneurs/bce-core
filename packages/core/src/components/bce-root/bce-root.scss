bce-root {
  @include bce-color-default(primary);

  /* Primary and accent colors */
  @include bce-color-define(#215475 #fff, primary);
  @include bce-color-define(#5c2a53 #fff, secondary);
  @include bce-color-define(#72222f #fff, tertiary);

  /* Semantic colors */
  @include bce-color-define(#1a562f #fff, success);
  @include bce-color-define(#b53131 #fff, error);
  @include bce-color-define(#d29e31 #fff, warning);
  @include bce-color-define(#3e8ebf #fff, info);

  /* Neutral colors */
  @include bce-color-define(#231f20 #fff, dark);
  @include bce-color-define(#f7f9f9 #000, light);
  @include bce-color-define(#cccccc #888, disable);
}

bce-root {
  position: relative;
  overflow-x: hidden;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica,
    Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol';

  * {
    box-sizing: border-box;
  }

  a {
    @include bce-color-default(primary);
    color: bce-color();

    &:hover {
      color: bce-color-l();
    }
  }

  ::selection {
    background-color: bce-color();
    color: bce-color-t();
  }

  &[fab] bce-message {
    right: 89px;
  }
}
