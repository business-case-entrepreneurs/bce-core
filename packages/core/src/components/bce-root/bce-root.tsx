import {
  Component,
  Element,
  h,
  Host,
  Method,
  Prop,
  State
} from '@stencil/core';

import { MessageOptions } from '../../models/message-options';
import { injectCSS } from '../../utils/stylesheet';
import { UUID } from '../../utils/uuid';

@Component({
  tag: 'bce-root',
  styleUrl: 'bce-root.scss',
  shadow: true
})
export class Root {
  @Element()
  private el!: HTMLBceRootElement;

  @Prop({ reflect: true })
  public mode?: 'default' | 'bucket';

  @State()
  private _fab = false;

  @State()
  private _messageQueue: ({ text: string } & MessageOptions)[] = [];

  private _messageTimer = 0;
  private _padding = 0;

  private handleSlotChange = (event: Event | HTMLSlotElement) => {
    const slot = 'target' in event ? (event.target as HTMLSlotElement) : event;
    if (!slot || slot.tagName !== 'SLOT') return;

    this._fab = !!this.el.querySelector('bce-fab:not([inline]');
    const header = this.el.querySelector('bce-header') ? 56 : 0;
    const status = this.el.querySelector('bce-status-bar') ? 20 : 0;
    const padding = header + status;

    if (this._padding !== padding) {
      this._padding = padding;

      const { id } = this.el;
      injectCSS(id, [
        `bce-root#${id} {`,
        `  --bce-padding-top: ${padding}px;`,
        '}'
      ]);
    }
  };

  @Method()
  public alert(title: string, message: string, options: AlertOptions = {}) {
    const dialog = document.createElement('bce-dialog');
    dialog.active = true;
    dialog.required = options.required == undefined ? true : options.required;
    dialog.innerHTML = `
      <h3>${title}</h3>
      <p>${message}</p>
    `;

    const action = document.createElement('bce-button');
    action.design = 'text';
    action.slot = 'action';
    action.type = 'submit';
    action.innerText = options.ok || 'Ok';
    if (options.ok !== false) dialog.appendChild(action);

    return this.execute<void>(
      dialog,
      res => {
        dialog.addEventListener('submit', () => res());
        dialog.addEventListener('backdrop', () => res());
        dialog.addEventListener('close', () => res());
      },
      { display: true }
    );
  }

  @Method()
  public confirm(title: string, message: string, options: ConfirmOptions = {}) {
    const dialog = document.createElement('bce-dialog');
    dialog.active = true;
    dialog.required = options.required == undefined ? true : options.required;
    dialog.innerHTML = `<h3>${title}</h3><p>${message}</p>`;

    const action2 = document.createElement('bce-button');
    action2.slot = 'action';
    action2.design = 'text';
    action2.style.opacity = '0.7';
    action2.innerText = options.cancel || 'Cancel';
    dialog.appendChild(action2);

    const action1 = document.createElement('bce-button');
    action1.slot = 'action';
    action1.innerText = options.ok || 'Ok';
    if (options.ok !== false) dialog.appendChild(action1);

    return this.execute<boolean>(
      dialog,
      res => {
        dialog.addEventListener('backdrop', () => res(false));
        action1.addEventListener('click', () => res(true));
        action2.addEventListener('click', () => res(false));
      },
      { display: true }
    );
  }

  @Method()
  public execute<T>(
    el: HTMLElement,
    exec: Executor<T>,
    options: ExecuteOptions = {}
  ): Promise<T> {
    const defaults: ExecuteOptions = { display: false };
    options = Object.assign({}, defaults, options);

    if (!options.display) el.style.display = 'none';

    const result = new Promise<T>((res, rej) => {
      const execution = exec(res, rej);
      if (execution) res(execution);
    });

    this.el.appendChild(el);
    result.then(() => this.el.removeChild(el));

    return result;
  }

  @Method()
  public async success(text: string, options: Partial<MessageOptions> = {}) {
    return this.message(text, { color: 'success', duration: 2, ...options });
  }

  @Method()
  public async error(text: string, options: Partial<MessageOptions> = {}) {
    return this.message(text, { color: 'error', duration: 5, ...options });
  }

  @Method()
  public async warning(text: string, options: Partial<MessageOptions> = {}) {
    return this.message(text, { color: 'warning', duration: 2, ...options });
  }

  @Method()
  public async info(text: string, options: Partial<MessageOptions> = {}) {
    return this.message(text, { color: 'info', duration: 2, ...options });
  }

  @Method()
  public async message(text: string, options: Partial<MessageOptions> = {}) {
    if (!text) return;

    // Messages have a minimum duration of 1 second and a maximum of 10 seconds
    let duration = options.duration || 2;
    duration = duration < 1 ? 1 : duration;
    duration = duration > 10 ? 10 : duration;

    const color = options.color || 'dark';
    this._messageQueue = [...this._messageQueue, { text, color, duration }];
    this.messageCheck();
  }

  private messageCheck() {
    // Skip if there is no message or already displaying a message
    const [message] = this._messageQueue;
    if (!message || !!this._messageTimer) return;

    // Remove message after specified duration
    this._messageTimer = window.setTimeout(
      () => this.messageTimeout(),
      message.duration * 1000
    );
  }

  private messageTimeout() {
    // Remove current message from queue
    this._messageQueue = this._messageQueue.slice(1);

    // Clear timer & check remaining queue
    this._messageTimer = 0;
    this.messageCheck();
  }

  componentDidLoad() {
    const slot = this.el.shadowRoot!.querySelector('slot');
    if (slot) {
      slot.addEventListener('slotchange', this.handleSlotChange);
      this.handleSlotChange(slot);
    }

    if (!this.el.id) this.el.id = UUID.v4();
  }

  renderMessage() {
    const [message] = this._messageQueue;
    if (!message) return;

    return (
      <bce-message
        key={this._messageTimer}
        color={message.color}
        fab={this._fab}
      >
        {message.text}
      </bce-message>
    );
  }

  render() {
    return (
      <Host>
        <slot />
        {this.renderMessage()}
      </Host>
    );
  }
}

export interface AlertOptions {
  readonly required?: boolean;
  readonly ok?: string | false;
}

export interface ConfirmOptions {
  readonly required?: boolean;
  readonly cancel?: string;
  readonly ok?: string | false;
}

export interface ExecuteOptions {
  readonly display?: boolean;
}

export type Executor<T> = (
  resolve: (value?: T | PromiseLike<T>) => void,
  reject: (reason?: any) => void
) => Promise<any> | void;
