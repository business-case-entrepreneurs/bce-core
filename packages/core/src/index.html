<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0"
    />
    <meta name="theme-color" content="#080707" />
    <title>BCE Core</title>

    <link rel="stylesheet" href="/build/bce.css" />
  </head>
  <body>
    <bce-root id="app">
      <bce-status-bar color="dark"></bce-status-bar>
      <bce-header color="dark">
        <bce-logo logo-src="/logo/logo.png"></bce-logo>
        <bce-menu-button color="light"></bce-menu-button>
      </bce-header>

      <bce-menu color="dark" verticale right>
        <a href="#switch">Switch</a>
        <a href="#input">Input</a>
        <a href="#button">Button</a>
        <a href="#message">Message</a>
      </bce-menu>

      <main>
        <h3 id="dropdown">BCE Dropdown menu</h3>
        <div>
          <bce-dropdown-menu>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
          </bce-dropdown-menu>
        </div>
        <h3 id="slider">BCE Slider</h3>
        <div><bce-slider max="200"></bce-slider></div>
        <h3 id="switch">BCE Switch</h3>
        <div>
          <bce-switch value="true"></bce-switch>
          <bce-switch value="true" color="secondary"></bce-switch>
          <bce-switch value="true" color="tertiary"></bce-switch>
        </div>
        <div>
          <bce-switch value="true" color="success"></bce-switch>
          <bce-switch value="true" color="error"></bce-switch>
          <bce-switch value="true" color="warning"></bce-switch>
          <bce-switch value="true" color="info"></bce-switch>
        </div>
        <div>
          <bce-switch value="true" color="dark"></bce-switch>
          <bce-switch value="true" color="light"></bce-switch>
        </div>

        <h3 id="input">BCE Input</h3>
        <bce-input
          label="Username"
          value="my-username"
          info="What's your name?"
        ></bce-input>
        <bce-input
          type="password"
          label="Password"
          info="This should be a value only you know..."
        ></bce-input>
        <bce-input
          type="number"
          label="OTP code"
          info="The One Time Password code."
          disabled
        ></bce-input>

        <bce-input type="dropdown" label="Dropdown" info="testing..">
          <option value="A">Option A</option>
          <option value="B">Option B</option>
          <option value="C">Option C</option>
          <option value="D">Option D</option>
          <option value="E">Option E</option>
          <option value="F">Option F</option>
          <option value="G">Option G</option>
        </bce-input>

        <h3 id="button">BCE Button</h3>
        <h4>Contained</h4>
        <div>
          <bce-button>Primary</bce-button>
          <bce-button color="secondary">Secondary</bce-button>
          <bce-button color="tertiary">Tertiary</bce-button>
          <bce-button icon="square">Icon</bce-button>
          <bce-button disabled>Disabled</bce-button>
        </div>
        <br />
        <h4>Outline</h4>
        <div>
          <bce-button type="outline">Primary</bce-button>
          <bce-button type="outline" color="secondary">Secondary</bce-button>
          <bce-button type="outline" color="tertiary">Tertiary</bce-button>
          <bce-button type="outline" icon="square">Icon</bce-button>
          <bce-button type="outline" disabled>Disabled</bce-button>
        </div>
        <br />
        <h4>Text</h4>
        <div>
          <bce-button type="text">Primary</bce-button>
          <bce-button type="text" color="secondary">Secondary</bce-button>
          <bce-button type="text" color="tertiary">Tertiary</bce-button>
          <bce-button type="text" icon="square">Icon</bce-button>
          <bce-button type="text" disabled>Disabled</bce-button>
        </div>
        <h4>Block</h4>
        <div>
          <bce-button color="tertiary" block>Contained</bce-button>
          <bce-button icon="square" block>Contained + Icon</bce-button>
          <bce-button type="outline" color="tertiary" block>Outline</bce-button>
          <bce-button type="outline" icon="square" block>
            Outline + Icon
          </bce-button>
          <bce-button type="text" color="tertiary" block>Text</bce-button>
          <bce-button type="text" icon="square" block>Text + Icon</bce-button>
        </div>

        <h3>BCE message</h3>
        <bce-button id="message">Trigger</bce-button>

        <h3 id="dropdown">BCE Dropdown menu</h3>
        <div>
          <bce-dropdown-menu>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
            <bce-button color="light" type="text" block>Primary</bce-button>
          </bce-dropdown-menu>
        </div>

        <h3>Heading 3</h3>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipisicing elit.
          Reprehenderit assumenda omnis sint, tenetur explicabo optio quam natus
          perferendis. Cupiditate perspiciatis ea, assumenda libero enim
          praesentium atque repellendus voluptatem explicabo perferendis.
        </p>
        <p>
          Fugiat praesentium distinctio officiis expedita consequuntur vitae
          amet ipsa enim odit nesciunt facere alias aliquid cupiditate
          perferendis, nobis aut qui velit maiores recusandae sed? Facilis
          aliquam rem autem! Dolores, modi?
        </p>
        <p>
          Sunt rem praesentium exercitationem eveniet excepturi impedit delectus
          qui similique quaerat! Fugit obcaecati sint quaerat animi cumque a
          beatae? Error, facilis necessitatibus? Provident, perferendis
          doloribus? Delectus quibusdam fugiat nulla vel!
        </p>
        <p>
          Quis delectus iure perferendis exercitationem suscipit necessitatibus
          doloribus? In dolore voluptates temporibus odit sapiente ea provident
          rerum, officia et magni consequuntur esse est unde, cumque accusamus
          dignissimos? Ullam, sit quidem?
        </p>
        <p>
          Aliquam odio excepturi quis reiciendis, recusandae dolor at, quam
          ipsam sunt magni modi beatae libero maiores similique amet iusto minus
          aspernatur quisquam. Veritatis, ducimus illum. Fugiat iure dolores
          assumenda voluptatum!
        </p>
        <p>
          Aperiam beatae quo cupiditate hic recusandae ad laudantium
          voluptatibus fugiat odit, harum sunt officia sed eaque unde ipsam
          obcaecati non magni voluptatem adipisci quia dignissimos dolores modi.
          Optio, tempora mollitia.
        </p>

        <bce-fab color="dark">
          <bce-button>TODO</bce-button>
          <bce-button>animate</bce-button>
          <bce-button>floating</bce-button>
          <bce-button>action</bce-button>
          <bce-button>button</bce-button>
          <bce-button>(de)activation</bce-button>
        </bce-fab>
      </main>
    </bce-root>

    <script type="module" src="build/bce.esm.js"></script>
    <script nomodule src="build/bce.js"></script>
    <script>
      var root = document.getElementById('app');
      var message = document.getElementById('message');

      var i = 0;
      message.addEventListener('click', function() {
        root.message('Message ' + ++i);
      });
    </script>
  </body>
</html>
