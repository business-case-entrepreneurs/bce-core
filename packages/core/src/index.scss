html,
body {
  margin: 0;
  height: 100%;
}

bce-root {
  height: 100vh;
  display: flex;
  flex-flow: column;

  // ::-webkit-scrollbar {
  //   background-color: #fff;
  //   width: 16px;
  // }
  //
  // ::-webkit-scrollbar-track {
  //   background-color: #fff;
  // }
  //
  // /* scrollbar itself */
  // ::-webkit-scrollbar-thumb {
  //   background-color: #babac0;
  //   border-radius: 16px;
  //   border: 4px solid #fff;
  // }
  //
  // ::-webkit-scrollbar-button {
  //   display: none;
  // }
}

bce-root bce-status-bar {
  flex: 0 0 auto;
}

bce-root bce-header {
  grid-column: span 3;
}

bce-root main {
  position: relative;
  padding: 1rem 1rem 82px;
  height: 100%;
  overflow-x: hidden;
}

@media (min-width: 768px) {
  bce-root {
    display: grid;
    grid-template-columns: 1fr auto;
    grid-template-rows: auto auto 1fr;
  }
}
