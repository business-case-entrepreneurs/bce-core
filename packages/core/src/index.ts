import { ColorScheme } from './models/color-scheme';
import { getColorShade, setColorShade } from './utils/color';
import { FileManager } from './utils/file-manager';
import { ScrollSpy } from './utils/scroll-spy';

export * from './models/color-scheme';
export * from './models/condition';
export * from './models/file-ref';
export * from './models/file-server';
export * from './models/file-store';
export * from './utils/bce-file';
export * from './utils/color';
export * from './utils/debounce';
export * from './utils/file-manager';
export * from './utils/is-email';
export * from './utils/modulo';
export * from './utils/ripple';
export * from './utils/scroll-spy';
export * from './utils/throttle';
export * from './utils/uuid';
export * from './utils/validator';

declare global {
  interface BceCore {
    file: FileManager;
    FileManager: typeof FileManager;
    generateId: () => string;
    getColorScheme: () => ColorScheme;
    getColorShade: typeof getColorShade;
    setColorScheme: (scheme: ColorScheme) => void;
    setColorShade: typeof setColorShade;
    ScrollSpy: typeof ScrollSpy;
  }

  interface Window {
    BCE: BceCore;
  }
}
