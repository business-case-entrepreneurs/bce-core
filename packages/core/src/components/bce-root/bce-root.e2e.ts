import { newE2EPage, E2EPage } from '@stencil/core/testing';

describe('Diff test', () => {
  it('render something', async () => {
    const page: E2EPage = await newE2EPage();
    await page.setContent('<my-cmp></my-cmp');
    await page.compareScreenshot(
      'My Componment (...is beautiful. Look at it!)',
      { fullPage: false }
    );
  });
});
