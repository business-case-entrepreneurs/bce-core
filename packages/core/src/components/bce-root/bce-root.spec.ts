import { newSpecPage } from '@stencil/core/testing';
import { Root } from './bce-root';

jest.mock('../../utils/uuid', () => {
  class UUID {
    public static v4(): string {
      return 'uidsgargrwthrt';
    }
  }
  return {
    UUID: UUID.v4()
  };
});

describe('testing bce-root', () => {
  it('builds', () => {
    expect(new Root()).toBeTruthy();
  });

  it('should render bce-root', async () => {
    const page = await newSpecPage({
      components: [Root],
      html: `<bce-root></bce-root>`
    });

    expect(page.root).toEqualHtml(`
    <bce-root id="agjkhgiureagerhgqh">
      <mock:shadow-root>
        <slot></slot>
      </mock:shadow-root>
    </bce-root>`);
  });
});
