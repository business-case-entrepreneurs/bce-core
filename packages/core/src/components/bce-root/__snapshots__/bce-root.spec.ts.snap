// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing bce-root bce-root should render message 1`] = `
<bce-root id="bce-root">
  <mock:shadow-root>
    <slot></slot>
    <bce-message color="dark">
      test
    </bce-message>
  </mock:shadow-root>
</bce-root>
`;

exports[`Testing bce-root render bce-root bce-root with unique id 1`] = `
<bce-root id="bce-root">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
</bce-root>
`;
